osu.Game/Graphics/UserInterface/FPSCounter.cs

```diff
@@ -0,0 +1,234 @@
+        {
+            background.FadeTo(idle_background_alpha, 200);
+            displayTemporarily();
+            base.OnHoverLost(e);
+        }
+
+        private bool isDisplayed;
+
+        private ScheduledDelegate? fadeOutDelegate;
+
+        private void displayTemporarily()
+        {
+            if (!isDisplayed)
+                mainContent.FadeTo(1, 300, Easing.OutQuint);
+
+            fadeOutDelegate?.Cancel();
+
+            if (!IsHovered)
+            {
+                fadeOutDelegate = Scheduler.AddDelayed(() =>
+                {
+                    mainContent.FadeTo(0, 300, Easing.OutQuint);
+                    isDisplayed = false;
+                }, 2000);
+            }
+        }
+
+        [Resolved]
+        private GameHost gameHost { get; set; } = null!;
+
+        protected override void Update()
+        {
+            base.Update();
+
+            // TODO: this is wrong (elapsed clock time, not actual run time).
+            double newFrameTime = gameHost.UpdateThread.Clock.ElapsedFrameTime;
+            double newFps = gameHost.DrawThread.Clock.FramesPerSecond;
+
+            bool hasSignificantChanges =
+                Math.Abs(msCounter.Current.Value - newFrameTime) > 5 ||
+                Math.Abs(fpsCounter.Current.Value - newFps) > 10;
+
+            if (hasSignificantChanges)
+                displayTemporarily();
+
+            // If the frame time spikes up, make sure it shows immediately on the counter.
+            if (msCounter.Current.Value < 20 && newFrameTime > 20)
+                msCounter.SetCountWithoutRolling(newFrameTime);
+            else
+                msCounter.Current.Value = newFrameTime;
+
+            fpsCounter.Current.Value = newFps;
+
+            fpsCounter.Colour = getColour(fpsCounter.DisplayedCount / gameHost.DrawThread.Clock.MaximumUpdateHz);
+
+            double equivalentHz = 1000 / msCounter.DisplayedCount;
+
+            msCounter.Colour = getColour(equivalentHz / gameHost.UpdateThread.Clock.MaximumUpdateHz);
+        }
+
+        private ColourInfo getColour(double performanceRatio)
+        {
+            if (performanceRatio < 0.5f)
+                return Interpolation.ValueAt(performanceRatio, colours.Red, colours.Orange2, 0, 0.5);
+
+            return Interpolation.ValueAt(performanceRatio, colours.Orange2, colours.Lime0, 0.5, 0.9);
+        }
+
+        public ITooltip GetCustomTooltip() => new FPSCounterTooltip();
+
+        public object TooltipContent => this;
+
+        public class FramesPerSecondCounter : RollingCounter<double>
+        {
+            protected override double RollingDuration => 400;
+
+            protected override OsuSpriteText CreateSpriteText()
+            {
+                return new OsuSpriteText
+                {
+                    Font = OsuFont.Default.With(fixedWidth: true, size: 16, weight: FontWeight.SemiBold),
+                    Spacing = new Vector2(-2),
+                };
```

Comment:
As for the counter alignment issue, I believe this is due to https://github.com/ppy/osu-framework/issues/3369, and fixing it would be to enforce children to have the same anchor/origin as the parent, i.e.:
```suggestion
                return new OsuSpriteText
                {
                    Anchor = Anchor.TopRight,
                    Origin = Anchor.TopRight,
                    Font = OsuFont.Default.With(fixedWidth: true, size: 16, weight: FontWeight.SemiBold),
                    Spacing = new Vector2(-2),
                };
```

Before:

https://user-images.githubusercontent.com/22781491/180103244-2b36d196-9a95-4d8d-b7ad-6fe0ca493668.mp4

After:

https://user-images.githubusercontent.com/22781491/180103247-55279363-4158-4f67-8cf6-254f3b537b1d.mp4

Reply:
Perfect.